contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,377 @@
+}
+
+```
+
+
+### CSI External Snapshot Controller
+* External snapshotter is responsible for creating/deleting snapshots and binding snapshot and snapshotData objects.
+
+* External snapshotter is running in the sidecar along with external-attacher and external-provisioner for each CSI Volume Driver.
+
+* For dynamically created snapshot, it should have a VolumeSnapshotClass associated with it. User can explicitly specify a VolumeSnapshotClass in the VolumeSnapshot API object. If user does not specify a VolumeSnapshotClass, a default VolumeSnapshotClass created by the admin will be used. This is similar to how a default StorageClass created by the admin will be used for the provisioning of a PersistentVolumeClaim.
+
+* For statically binding snapshot, user/admin must specify pointers for both VolumeSnapshot and VolumeSnapshotContent, so that the controller knows how to bind them.
+
+![CSI Snapshot Diagram](csi-snapshot_diagram.png?raw=true "CSI Snapshot Diagram")
+
+
+External snapshotter is part of Kubernetes implementation of [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec). It is an external controller that monitors `VolumeSnapshot` and `VolumeSnapshotContent` objects and creates/deletes snapshot from volumes. The original Kubernetes CSI design can be found at Kubernetes [design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md).
+
+External snapshotter follows [controller](https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md) pattern and uses informers to watch for `VolumeSnapshot` and `VolumeSnapshotContent` create/update/delete events. It filters out `VolumeSnapshot` instances with `Snapshotter==<CSI driver name>` and processes these events in workqueues with exponential backoff. Real handling is deferred to the `Handler` interface.
+
+The snapshotter talks to out-of-tree CSI Volume Driver over socket (`/run/csi/socket` by default, configurable by `-csi-address`). The snapshotter tries to connect for `-connection-timeout` (1 minute by default), allowing CSI Volume Driver to start and create its server socket a bit later.
+
+The snapshotter creates CreateSnapshotRequest and calls CreateSnapshot through CSI ControllerClient. It gets CreateSnapshotResponse from the out-of-tree CSI Volume Driver and creates a VolumeSnapshotContent API object with VolumeSnapshotSource.
+
+The snapshotter checks status in VolumeSnapshot to decide whether to bind. It binds VolumeSnapshot and VolumeSnapshotContent when ready.
+
+When the storage system fails to create snapshot, retry will not be performed in the first version. This is because users may not want to retry when taking consistent snapshots or scheduled snapshots when the timing of the snapshot creation is important. In a future version, a maxRetries flag will be added to allow users to control whether retries are needed.
```

Comment:
@garthy had a good recommendation on the call: should have retry termination timestamp instead of number of retries. Which will work better with workload "quiecing".

Reply:
Done.